File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/C Programs/5_count_nodes.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node { int data; struct Node* next; } Node;
Node* cn(int d) { Node* n = malloc(sizeof(Node)); n->data = d; n->next = NULL; return n; }
Node* ial(Node* h, int d) { Node* n = cn(d); if (!h) return n; Node* t = h; while (t->next) t = t->next; t->next = n; return h; }
int count(Node* h) { int c = 0; while (h) c++, h = h->next; return c; }
int countr(Node* h) { return h ? 1 + countr(h->next) : 0; }
int main() {
    Node* h = NULL; int d, c;
    int vals[] = {10, 20, 30, 40, 50};
    for (int i = 0; i < 5; i++) h = ial(h, vals[i]);
    while (scanf("%d", &c) && c != 5) {
        if (c == 1) scanf("%d", &d), h = ial(h, d);
        else if (c == 2) printf("%d\n", count(h));
        else if (c == 3) printf("%d\n", countr(h));
    }
    return 0;
}

File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/Python Programs/5_count_nodes.py
class Node:
    def __init__(self, d): self.data, self.next = d, None

class LL:
    def __init__(self): self.head = None
    def add(self, d): n = Node(d); (setattr(self, 'head', n) if not self.head else [t for t in [self.head] if not [setattr(t, 'next', n) for t in iter(lambda x=self.head: x.next if x.next else None, None)]])
    def count(self): c, t = 0, self.head; [(c := c + 1, t := t.next) for _ in iter(lambda: t, None)]; return c
    def countr(self, n=None): n = n or self.head; return 0 if not n else 1 + self.countr(n.next)

ll = LL()
[ll.add(i) for i in [10, 20, 30, 40, 50]]
print(f"Iterative: {ll.count()}")
print(f"Recursive: {ll.countr()}")